### PR TITLE
Interpolate multifield

### PR DESCRIPTION
**Pull request by oriolcg**

### CODE REVIEW

File: src/FESpaces/FESpacesWithLastDofRemoved.jl [hunk 1]
```diff
@@ -54,9 +54,11 @@ function scatter_free_and_dirichlet_values(
...
 end
 
-function gather_free_and_dirichlet_values(
-  f::FESpaceWithLastDofRemoved,cv)
-  _fv, _dv = gather_free_and_dirichlet_values(f.space,cv)
+function gather_free_and_dirichlet_values!(
+  fv,dv,f::FESpaceWithLastDofRemoved,cv)
+  _fv = zero_free_values(f.space)
+  _dv = zero_dirichlet_values(f.space)
+  gather_free_and_dirichlet_values!(_fv,_dv,f.space,cv)
```
Comment (fverdugo) [inline src/FESpaces/FESpacesWithLastDofRemoved.jl:61]: For efficiency reasons, keep the original function
```julia
function gather_free_and_dirichlet_values(
  f::FESpaceWithLastDofRemoved,cv)
  _fv, _dv = gather_free_and_dirichlet_values(f.space,cv)
  @assert length(_dv) == 0
  l = length(_fv)
  fv = SubVector(_fv,1,l-1)
  dv = SubVector(_fv,l,l)
  (fv, dv)
end
```
and implement the new one separatelly (note the `.=` in the last lines )
```julia
function gather_free_and_dirichlet_values!(
  fv,dv,f::FESpaceWithLastDofRemoved,cv)
  _fv, _dv = gather_free_and_dirichlet_values(f.space,cv)
  @assert length(_dv) == 0
  l = length(_fv)
  fv .= SubVector(_fv,1,l-1)
  dv .= SubVector(_fv,l,l)
  (fv, dv)
end
```

File: src/FESpaces/FESpacesWithLinearConstraints.jl [hunk 2]
```diff
@@ -404,10 +404,10 @@ function _setup_dof_to_val!(
...
 end
 
-function gather_free_and_dirichlet_values(f::FESpaceWithLinearConstraints,cell_to_ludof_to_val)
-  fdof_to_val, ddof_to_val = gather_free_and_dirichlet_values(f.space,cell_to_ludof_to_val)
-  fmdof_to_val = zero_free_values(f)
-  dmdof_to_val = zero_dirichlet_values(f)
+function gather_free_and_dirichlet_values!(fmdof_to_val,dmdof_to_val,f::FESpaceWithLinearConstraints,cell_to_ludof_to_val)
+  fdof_to_val = zero_free_values(f.space)
+  ddof_to_val = zero_dirichlet_values(f.space)
+  gather_free_and_dirichlet_values!(fdof_to_val, ddof_to_val,f.space,cell_to_ludof_to_val)
```
Comment (fverdugo) [inline src/FESpaces/FESpacesWithLinearConstraints.jl:410]: Idem here. Keep the original one

```julia
function gather_free_and_dirichlet_values(f::FESpaceWithLinearConstraints,cell_to_ludof_to_val)
  fdof_to_val, ddof_to_val = gather_free_and_dirichlet_values(f.space,cell_to_ludof_to_val)
  fmdof_to_val = zero_free_values(f)
  dmdof_to_val = zero_dirichlet_values(f)
  _setup_mdof_to_val!(
    fmdof_to_val,
    dmdof_to_val,
    fdof_to_val,
    ddof_to_val,
    f.mDOF_to_DOF,
    f.n_fdofs,
    f.n_fmdofs)
  fmdof_to_val, dmdof_to_val
end
```
and add the now one separatelly

```julia
function gather_free_and_dirichlet_values (
  fmdof_to_val,dmdof_to_val,f::FESpaceWithLinearConstraints,cell_to_ludof_to_val)

  fdof_to_val, ddof_to_val = gather_free_and_dirichlet_values(f.space,cell_to_ludof_to_val)
  _setup_mdof_to_val!(
    fmdof_to_val,
    dmdof_to_val,
    fdof_to_val,
    ddof_to_val,
    f.mDOF_to_DOF,
    f.n_fdofs,
    f.n_fmdofs)
  fmdof_to_val, dmdof_to_val
end
```

File: src/FESpaces/SingleFieldFESpaces.jl [hunk 3]
```diff
@@ -64,6 +63,7 @@ function test_single_field_fe_space(f::SingleFieldFESpace,pred=(==))
   free_values = zero_free_values(f)
   cell_vals = scatter_free_and_dirichlet_values(f,free_values,dirichlet_values)
   fv, dv = gather_free_and_dirichlet_values(f,cell_vals)
+  gather_free_and_dirichlet_values!(fv,dv,f,cell_vals)
```
Comment (fverdugo) [inline src/FESpaces/SingleFieldFESpaces.jl:66]: add this minor fix (safety reasons)

```julia
  fv, dv = gather_free_and_dirichlet_values(f,cell_vals)
  @test pred(fv,free_values)
  @test pred(dv,dirichlet_values)

  gather_free_and_dirichlet_values!(fv,dv,f,cell_vals)
  @test pred(fv,free_values)
  @test pred(dv,dirichlet_values)

  fv, dv = gather_free_and_dirichlet_values!(fv,dv,f,cell_vals)
  @test pred(fv,free_values)
  @test pred(dv,dirichlet_values)
```

File: src/FESpaces/SingleFieldFESpaces.jl [hunk 4]
```diff
@@ -273,16 +306,32 @@ like interpolate, but also compute new degrees of freedom for the dirichlet comp
...
+  interpolate_dirichlet!(free_values,dirichlet_values,fs,object)
+end
+
+"""
+"""
+function interpolate_dirichlet!(free_values,dirichlet_values,fs::SingleFieldFESpace,object)
   cell_vals = _cell_vals(fs,object)
-  dirichlet_values = gather_dirichlet_values(fs,cell_vals)
+  gather_dirichlet_values!(dirichlet_values,fs,cell_vals)
   free_values = zero_free_values(fs)
```
Comment (fverdugo) [inline src/FESpaces/SingleFieldFESpaces.jl:335]: Replace 
```julia
free_values = zero_free_values(fs)
```
by
```julia
fill!(free_values,zero(eltype(free_values)))
```

File: src/FESpaces/TrialFESpaces.jl [hunk 5]
```diff
@@ -92,11 +92,11 @@ get_dirichlet_dof_tag(f::TrialFESpace) = get_dirichlet_dof_tag(f.space)
 
 scatter_free_and_dirichlet_values(f::TrialFESpace,fv,dv) = scatter_free_and_dirichlet_values(f.space,fv,dv)
 
-gather_free_and_dirichlet_values(f::TrialFESpace,cv) = gather_free_and_dirichlet_values(f.space,cv)
```
Comment (fverdugo) [inline src/FESpaces/TrialFESpaces.jl:95]: keep both the original and new functions

File: src/FESpaces/TrialFESpaces.jl [hunk 6]
```diff
@@ -92,11 +92,11 @@ get_dirichlet_dof_tag(f::TrialFESpace) = get_dirichlet_dof_tag(f.space)
...
+gather_free_and_dirichlet_values!(fv,dv,f::TrialFESpace,cv) = gather_free_and_dirichlet_values!(fv,dv,f.space,cv)
 
-gather_dirichlet_values(f::TrialFESpace,cv) = gather_dirichlet_values(f.space,cv)
```
Comment (fverdugo) [inline src/FESpaces/TrialFESpaces.jl:97]: keep both the original and new functions

File: src/FESpaces/TrialFESpaces.jl [hunk 7]
```diff
@@ -92,11 +92,11 @@ get_dirichlet_dof_tag(f::TrialFESpace) = get_dirichlet_dof_tag(f.space)
...
+gather_dirichlet_values!(dv,f::TrialFESpace,cv) = gather_dirichlet_values!(dv,f.space,cv)
 
-gather_free_values(f::TrialFESpace,cv) = gather_free_values(f.space,cv)
```
Comment (fverdugo) [inline src/FESpaces/TrialFESpaces.jl:99]: keep both the original and new functions

File: src/FESpaces/ZeroMeanFESpaces.jl [hunk 8]
```diff
@@ -110,9 +110,9 @@ get_dirichlet_dof_tag(f::ZeroMeanFESpace) = get_dirichlet_dof_tag(f.space)
 
 scatter_free_and_dirichlet_values(f::ZeroMeanFESpace,fv,dv) = scatter_free_and_dirichlet_values(f.space,fv,dv)
 
-gather_free_and_dirichlet_values(f::ZeroMeanFESpace,cv) = gather_free_and_dirichlet_values(f.space,cv)
```
Comment (fverdugo) [inline src/FESpaces/ZeroMeanFESpaces.jl:113]: keep both the original and new functions

File: src/FESpaces/ZeroMeanFESpaces.jl [hunk 9]
```diff
@@ -110,9 +110,9 @@ get_dirichlet_dof_tag(f::ZeroMeanFESpace) = get_dirichlet_dof_tag(f.space)
...
+gather_free_and_dirichlet_values!(fv,dv,f::ZeroMeanFESpace,cv) = gather_free_and_dirichlet_values!(fv,dv,f.space,cv)
 
-gather_dirichlet_values(f::ZeroMeanFESpace,cv) = gather_dirichlet_values(f.space,cv)
```
Comment (fverdugo) [inline src/FESpaces/ZeroMeanFESpaces.jl:115]: keep both the original and new functions

File: src/FESpaces/ZeroMeanFESpaces.jl [hunk 10]
```diff
@@ -110,9 +110,9 @@ get_dirichlet_dof_tag(f::ZeroMeanFESpace) = get_dirichlet_dof_tag(f.space)
...
+gather_dirichlet_values!(dv,f::ZeroMeanFESpace,cv) = gather_dirichlet_values!(dv,f.space,cv)
 
-gather_free_values(f::ZeroMeanFESpace,cv) = gather_free_values(f.space,cv)
```
Comment (fverdugo) [inline src/FESpaces/ZeroMeanFESpaces.jl:117]: keep both the original and new functions

**Comment by oriolcg** [comment]
Requested modifications added in the recent commit.

**Pull request merged**